### DISCUSSION TITLE
SPM build tool plugin support

### DISCUSSION THREAD

**Discussion by vladvlasov256** [mono0926/LicensePlist#209] [Ideas]
# Build Tools

As you probably know, Xcode 14 supports [SPM build tool plugins](https://github.com/apple/swift-evolution/blob/main/proposals/0303-swiftpm-extensible-build-tools.md). That means tools like [mint](https://github.com/yonaskolb/Mint) could be replaced with the plugins.

# LicensePlist Plugin

Today I was able to generate `Acknowledgements.plist` with a build plugin using patched `LicensePlist`. All related work is in [this branch](https://github.com/vladvlasov256/LicensePlist/tree/feature/build_tool).

I haven't updated the readme and tests yet but the plugin already could be used in Xcode.

<img width="404" alt="Screenshot 2023-02-12 at 15 52 21" src="https://user-images.githubusercontent.com/1046502/218318301-3ed70e0b-27a6-442f-9962-531ea5216109.png">

# Limitations

## No Command-line Arguments

Build tool commands don't support command-line arguments. Thus all the configuration should be done in the config file. For this purpose, the plugin branch is based on another [pull request](https://github.com/mono0926/LicensePlist/pull/207)'s branch. 

## Write Access

Another limitation is that Xcode runs plugins in a sandbox. Plugin tools can read all project files but they can write only to their own output folders. At the same moment, Xcode automatically copies generated files to a build folder inside `Derived Data`. But that's not possible to simply generate `Settings.bundle` directory because it's a kind of duplication error for Xcode. It has one `Settings.bundle` linked from project files and one copied by the plugin. Unfortunately, Xcode can't merge them automatically.

As a workaround I added to my test project a script that manually copies the result files:
```bash
echo "Will copy Acknowledgements.plist"

ACKNOWLEDGEMENTS_DIR=${BUILT_PRODUCTS_DIR}/${CONTENTS_FOLDER_PATH}/LicensePlist
ACKNOWLEDGEMENTS_PLIST=${ACKNOWLEDGEMENTS_DIR}/Acknowledgements.plist
SETTINGS_BUNDLE=${BUILT_PRODUCTS_DIR}/${CONTENTS_FOLDER_PATH}/Settings.bundle/

cp ${ACKNOWLEDGEMENTS_PLIST} ${SETTINGS_BUNDLE}
rm -rf ${ACKNOWLEDGEMENTS_DIR}
```

`LicensePlist` here is a hardcoded name of the output folder. This script supports only single-paged acknowledgments. But it's not hard to support multiple copying.

## No Network Access

Another sandbox limitation is network connection restrictions. For GitHub packages, `LicensePlist` fetches licenses via GitHub API. Which is impossible to do in the sandbox. As a workaround, I used a path to `checkout` folder inside `Derived Data`. During package resolving Xcode fetches the sources. Thus the tool can read `LICENSE` files from disk instead of API calls.

# Changes

What's changed in the tool:
- added plugin target to package definition,
- added binary target to package definition,
- added `buildTool` parameter to let the plugin explicitly specify that the tool is used in a build tool mode,
- added `packageCheckoutPath` parameter with a path to package source codes,
- added a method to read GitHub library licenses from disk,
- added a method to read package definition from disk for default name resolution (`nameSpecified`).

# Future work

## Binary Target

Makefile should be updated to support binary targets. I'm not sure that it's an Xcode bug. But currently, Xcode doesn't support executable targets for prebuild commands. As a workaround plugin uses downloaded binaries which I manually pushed to my repo.

## SPM Command Plugin

I didn't have time for implementing a command plugin but I think it could be a helpful feature. Commands grant explicit write access. No need to have any scripts to copy result files. At the same moment, commands can't be called automatically. I'd prefer an automatic tool rather than a command in a context menu.

----

If @mono0926 thinks the path I chose is the right one I provide a pull request for this feature.

And I'd like to hear any comments regarding the implementation.

**Comment by mono0926** [comment]
@vladvlasov256 

Thank you for the nice proposal 👍 

> If @mono0926 thinks the path I chose is the right one I provide a pull request for this feature.

I'm welcome your pull request 🙏

**Comment by vladvlasov256** [comment]
[The 2nd](https://github.com/mono0926/LicensePlist/pull/210) of 4 planned PRs is ready.

The PR adds a binary target to the package. But mostly it's about automating artifact uploading for that target.

It's <s>stolen from</s> inspired by how [SwiftLint](https://github.com/realm/SwiftLint) organizes stuff.

**Comment by vladvlasov256** [comment]
Current progress:
- the tool has a binary target (#210) and it supports "offline" work with SPM (#211 🎉
- the next step is providing a plugin 💪

**Reply by Patrick-Kladek** [reply]
Hi @vladvlasov256 any update on this? This would be really useful and I could help if needed.

**Reply by vladvlasov256** [reply]
Hi @Patrick-Kladek! My bad, I should close this discussion. The plugin support (#212) were merged in April, 2023.
